File: backend/src/services/diagnosis.ts
import { analyzeCrop } from "./ai.service.js";
import { Diagnosis } from "../models/Diagnosis.js";

export async function analyzeCropImage(
  crop: string,
  imageUrl: string
) {
  const diagnosis = await analyzeCrop({
    crop,
    image_url: imageUrl,
  });

  return diagnosis;
}




export async function saveDiagnosis(
  diagnosis: Awaited<ReturnType<typeof analyzeCrop>>,
  imageUrl: string,
  imageKey?: string,
  userId?: string
) {
  const savedDiagnosis = await Diagnosis.create({
    userId: userId || undefined,

    crop: diagnosis.crop,

    imageUrl,

    imageKey,

    healthStatus: diagnosis.health_status,

    disease: {
      name: diagnosis.disease.name,
      confidence: diagnosis.disease.confidence,
    },

    observations: diagnosis.observations,

    agronomistAdvice: {
      overview: diagnosis.agronomist_advice.overview,
      symptoms: diagnosis.agronomist_advice.symptoms,
      causes: diagnosis.agronomist_advice.causes,
      treatment: diagnosis.agronomist_advice.treatment,
      prevention: diagnosis.agronomist_advice.prevention,
      severity: diagnosis.agronomist_advice.severity,
      recommendedAction:
        diagnosis.agronomist_advice.recommended_action,
    },
  });

  return savedDiagnosis;
}

File: ai_service/app/schemas/diagnosis.py
from typing import Optional

from pydantic import BaseModel, Field, field_validator


class DiseaseDetectionRequest(BaseModel):
    crop: str = Field(
        ...,
        min_length=2,
        max_length=50,
        description="Crop being analyzed",
    )

    image_url: str = Field(
        ...,
        min_length=10,
        description="URL of the crop image",
    )

    @field_validator("crop")
    @classmethod
    def validate_crop(cls, value: str) -> str:
        return value.strip().lower()


class DiseaseInfo(BaseModel):
    name: Optional[str] = None

    confidence: float = Field(
        ...,
        ge=0.0,
        le=1.0,
    )


class DiseaseDetectionResponse(BaseModel):
    crop: str
    health_status: str
    disease: DiseaseInfo
    observations: list[str]


class AgronomistAdvice(BaseModel):
    overview: str
    symptoms: list[str]
    causes: list[str]
    treatment: list[str]
    prevention: list[str]
    severity: str
    recommended_action: str


class FullDiagnosisResponse(BaseModel):
    crop: str
    health_status: str
    disease: DiseaseInfo
    observations: list[str]
    agronomist_advice: AgronomistAdvice